### PR TITLE
Fix: fencer,libstonithd: add `port` or `plug` parameter according to metadata for RHCS-style fence-agents

### PR DESCRIPTION
**Pull request by gao-yan**
This addresses the issue brought up from: https://github.com/ClusterLabs/pacemaker/pull/2036#issuecomment-621529506

According to:
https://github.com/ClusterLabs/fence-agents/commit/de490e059

, `port` parameter is massively deprecated in favor of `plug`:

```
<parameter name="plug" ... obsoletes="port">...
<parameter name="port" ... deprecated="1">...
```

With this commit, according to the metadata, if `port` parameter is not
supported, `plug` parameter is added if supported. By default, `port`
is added.

### CODE REVIEW

**Comment by gao-yan** [comment]
Probably let's wait for a good solution for stonith__rhcs_validate() to be appended here.

File: daemons/fenced/fenced_commands.c [hunk 1]
```diff
@@ -407,11 +408,19 @@ stonith_device_execute(stonith_device_t * device)
         action_str = "off";
     }
 
+    if (is_set(device->flags, st_device_supports_parameter_port)) {
+        host_arg = "port";
+
+    } else if (is_set(device->flags, st_device_supports_parameter_plug)) {
+        host_arg = "plug";
+    }
```
Comment (kgaillot) [inline daemons/fenced/fenced_commands.c:416]: If both are specified, should we prefer port or plug?

**Comment by gao-yan** [comment]
> Looks good so far.
> 
> For validation, one issue is that we strip out pcmk_* in stonith_api_validate(). We should probably look for pcmk_host_argument specially and pass it to stonith__rhcs_validate(). Only if that is NULL do we have to worry about meta-data.
> 
> I suppose it does make sense to run meta-data in that case. I'm a little uncomfortable with meta-data eating into the timeout, but the only current usage of validate is in the stonith_admin command line so it's not horrible. Also we currently ignore the timeout in stonith__rhcs_validate() so it's rather moot at the moment, though that should be fixed at some point.

Agreed.

> 
> Don't forget to update pcmk_host_argument in Pacemaker Explained.

You mean the specialty when it's supplied for `stonith_admin --validate`?

File: lib/fencing/st_client.c [hunk 2]
```diff
@@ -550,7 +551,14 @@ make_args(const char *agent, const char *action, const char *victim, uint32_t vi
             value = agent;
 
         } else if (param == NULL) {
-            param = "port";
+            // By default, `port` is added
+            if (host_arg == NULL) {
+                param = "port";
```
[the remark this answers is not in the document]
Comment (gao-yan) [reply lib/fencing/st_client.c:556]: Yes, makes sense. That'll be the deadline for any problematic fence agents to fix their potential issues.

File: daemons/fenced/fenced_commands.c (hunk 1 above)
Comment (gao-yan) [reply daemons/fenced/fenced_commands.c:416]: I'm kind of hesitant too... "port" for the best backward compatibility as how it is now, "plug" for the favor of the new standard though ...

**Comment by kgaillot** [comment]
> > Don't forget to update pcmk_host_argument in Pacemaker Explained.
> 
> You mean the specialty when it's supplied for `stonith_admin --validate`?

The default/description just need to be updated for this PR's changes

**Comment by gao-yan** [comment]
> > For validation, one issue is that we strip out pcmk_* in stonith_api_validate(). We should probably look for pcmk_host_argument specially and pass it to stonith__rhcs_validate(). Only if that is NULL do we have to worry about meta-data.

Don't know how far `validate-all` action of fence agents will go. Does it make sense to as well parse and handle `pcmk_host_map`  here?

**Comment by kgaillot** [comment]
> Don't know how far `validate-all` action of fence agents will go. Does it make sense to as well parse and handle `pcmk_host_map` here?

No, we intentionally use a generic "node1" target name with validate-all, we just need to provide it using a parameter it understands. That works fine with existing fence agents. An agent could theoretically try to verify that it can fence a specific host, but in practice validate-all is just used to make sure the parameters make sense, and the target needs to be specified only because most agents mark it as a required parameter.

**Comment by kgaillot** [comment]
Looks good to me, ready to merge?

**Comment by gao-yan** [comment]
Let's wait. Improvement for stonith__rhcs_validate() is coming soon :-)

**Comment by kgaillot** [comment]
Whoops, I remembered talking about it but forgot it hadn't been done yet :)

**Comment by kgaillot** [comment]
Thanks!

**Pull request merged**